File: lintcode/edit-distance.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/edit-distance/

class Solution:
    # @param word1 & word2: Two string.
    # @return: The minimum number of steps.
    def minDistance(self, word1, word2):
        # write your code here
        if len(word1) == 0 or len(word2) == 0:
            return abs(len(word1) - len(word2))

        row = len(word1) + 1
        col = len(word2) + 1

        dp = [[0] * col for i in xrange(row)]
        for i in xrange(1, row):
            dp[i][0] = i
        for j in xrange(1, col):
            dp[0][j] = j

        for i in xrange(1, row):
            for j in xrange(1, col):
                if word1[i-1] == word2[j-1]:
                    dp[i][j] = min(dp[i-1][j-1], dp[i-1][j] + 1, dp[i][j-1] + 1)
                else:
                    dp[i][j] = min(dp[i-1][j-1], dp[i-1][j], dp[i][j-1]) + 1

        return dp[row-1][col-1]

if __name__ == '__main__':
    s = Solution()
    print s.minDistance('mart', 'karma')
    print s.minDistance('a', 'a')

File: leetcode-2/edit-distance.cc
// Question: https://leetcode.com/problems/edit-distance/

class Solution {
public:
    int minDistance(string word1, string word2) {
        int row = word1.size();
        int col = word2.size();
        vector<vector<int>> dp(row+1, vector<int>(col+1, 0));
        for (int i = 1; i <= row; ++i) {
            dp[i][0] = i;
        }
        for (int j = 1; j <= col; ++j) {
            dp[0][j] = j;
        }

        for (int i = 0; i < row; ++i) {
            for (int j = 0; j < col; ++j) {
                if (word1[i] == word2[j]) {
                    dp[i+1][j+1] = min(dp[i][j], min(dp[i][j+1], dp[i+1][j]) + 1);
                } else {
                    dp[i+1][j+1] = min(dp[i][j], min(dp[i][j+1], dp[i+1][j])) + 1;
                }
            }
        }

        return dp.back().back();
    }
};
